### DISCUSSION TITLE
Extracting UrlParams

### DISCUSSION THREAD

**Discussion by nyarly** [tokio-rs/axum#2856] [Q&A] [answered]
### Summary

I'm experimenting with an authorization framework, and I'd like to be able to extract facts about a general request in order to apply policy to it. Being able to consider the path params of a request would be very helpful.

I see https://github.com/tokio-rs/axum/pull/154, where the UrlParams extractor was removed, and that the UrlParams extension has been made `crate(public)`, but the reasoning behind that change isn't clear. The PR body says "use extract::Path" instead, but to do that I need to know how many parameters are in the path - which is sensible for the author of the Router and Handler, but as a middleware author that doesn't make sense.

Is there a different approach for extracting some unknown number of path parameters, perhaps equivalent to how the `Query` extractor can fill a `Vec`?

### axum version

0.7.5

**Comment by nyarly** [comment, marked as answer]
Aha - just found [`extract::RawPathParams`](https://docs.rs/axum/0.7/axum/extract/struct.RawPathParams.html) that seems to do what I need.

**Reply by jplatte** [reply]
Right, I always forget that's a thing ^^
That should be more efficient than my first two suggestions.

**Comment by jplatte** [comment]
Yes, you can extract all path parameters using `Path<Vec<String>>` if you just want the values, `Path<BTreeMap<String, String>>` (or `HashMap`) if you want a map of parameter names and values. You can even do

```rust
#[derive(Deserialize)]
struct Parameters {
    foo_id: Option<Uuid>,
    bar_id: Option<Uuid>,
    name: Option<String>,
}

// you can use `Path(params): Path<Parameters>` and if the route has a `foo_id` param,
// `params.foo_id` will have the deserialized parameter value
```
